anesthetic/labelled_pandas.py

```diff
@@ -0,0 +1,216 @@
+
+    Given a list of callables `funcs`, and their arguments `*args`, evaluate
+    each of these, catching exceptions, and then sort results by their
+    dimensionality, smallest first. Return the non-exceptional result with the
+    smallest dimensionality.
+    """
+    results = []
+    errors = []
+    for f in funcs:
+        try:
+            results.append(f(*args))
+        except Exception as e:
+            errors.append(e)
+
+    def cmp(x, y):
+        if x.ndim > y.ndim:
+            return 1
+        elif x.ndim < y.ndim:
+            return -1
+        else:
+            x_levels = 0
+            y_levels = 0
+            if x.ndim > 0:
+                x_levels += x.index.nlevels
+                y_levels += y.index.nlevels
+            if x.ndim > 1:
+                x_levels += x.columns.nlevels
+                y_levels += y.columns.nlevels
+
+            if x_levels < y_levels:
+                return 1
+            elif x_levels > y_levels:
+                return -1
+            else:
+                return 0
+
+    results.sort(key=cmp_to_key(cmp))
+
+    for s in results:
+        if s is not None:
+            return s
+    raise errors[-1]
+
+
+class _LocIndexer(_LocIndexer_):
+    def __getitem__(self, key):
+        return ac([_LocIndexer_("loc", self.obj.drop_labels(i)).__getitem__
+                   for i in self.obj._all_axes()] + [super().__getitem__], key)
+
+
+class _AtIndexer(_AtIndexer_):
+    def __getitem__(self, key):
+        return ac([_AtIndexer_("at", self.obj.drop_labels(i)).__getitem__
+                   for i in self.obj._all_axes()] + [super().__getitem__], key)
+
+
+class _LabelledObject(object):
+    """Common methods for LabelledSeries and LabelledDataFrame."""
+
+    def __init__(self, *args, **kwargs):
+        self._labels = ("labels", "labels")
+        labels = kwargs.pop(self._labels[0], None)
+        super().__init__(*args, **kwargs)
+        if labels is not None:
+            self.set_labels(labels, inplace=True)
+
+    def islabelled(self, axis=0):
+        """Determine if labels are actually present."""
+        return (self._labels[axis] is not None
+                and self._labels[axis] in self._get_axis(axis).names)
+
+    def get_labels(self, axis=0):
+        """Retrieve labels from an axis."""
+        if self.islabelled(axis):
+            return self._get_axis(axis).get_level_values(
+                    self._labels[axis]).to_numpy()
+        else:
+            return None
+
+    def get_labels_map(self, axis=0):
+        """Retrieve mapping from paramnames to labels from an axis."""
+        index = self._get_axis(axis)
+        if self.islabelled(axis):
+            return index.to_frame().droplevel('labels')['labels']
+        else:
+            return Series('', index=index)
+
+    def get_label(self, param, axis=0):
+        """Retrieve mapping from paramnames to labels from an axis."""
+        return self.get_labels_map(axis)[param]
```

Comment:
If we want to treat weights as a (special) label as well, then we need this distinction. I agree however that the current setup of `.set_label('x0', '$x_0$', axis=1, inplace=False` is too verbose, but it is consistent with pandas. We could perhaps have a shortcut which auto-detected the axis?

Reply:
We already have `samples.get_weights()`, so having `samples.get_labels()` return the same thing is a bit counter intuitive for me. 

Currently `Samples` is a child of `WeightedDataFrame` and `LabelledDataFrame`. Would it make more sense to have `WeightedDataFrame` be a child of `LabelledDataFrame` and `Samples` of `WeightedDataFrame` only? Then we could justify that `LabelledDataFrame` has `axis=0` as a default in line with pandas, but overwrite that default behaviour for `WeightedDataFrame`, where weights are a special label and get their own setters and getters, and therefore the default for labels can be set to `axis=1`...

Reply:
I quite like this proposal -- certainly neatens up the diamond inheritance structure, although it forces WeightedDataFrame to be labelled. I could put the same set of default fiddling in the `WeightedLabelledDataFrame`.

What about `inplace=True`? I'm torn here, because 'go with pandas' would definitely say `False` by default, but I can imagine users being (pandas) infuriated by trying `samples.set_label('x0', '$x_0$') and finding nothing happening. We could just simply not allow an `inplace` option for `set_label` (but retain it for the more wholesale `set_labels`).

Reply:
Yes to all of that.

Reply:
OK, implemented in 0bb2275 -- very satisfying vimming through and removing all the `axis=1` commands.